server/src/main/java/org/elasticsearch/repositories/ShardGenerations.java

```diff
@@ -0,0 +1,168 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.repositories;
+
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.xcontent.ToXContent;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public final class ShardGenerations implements ToXContent {
+
+    public static final ShardGenerations EMPTY = ShardGenerations.builder().build();
+
+    private final Map<IndexId, List<String>> shardGenerations;
+
+    private ShardGenerations(Map<IndexId, List<String>> shardGenerations) {
+        this.shardGenerations = shardGenerations;
+    }
+
+    /**
+     * Returns all indices for which shard generations are tracked.
+     *
+     * @return indices for which shard generations are tracked
+     */
+    public List<IndexId> indices() {
+        return List.copyOf(shardGenerations.keySet());
+    }
+
+    /**
+     * Computes the obsolete shard index generations that can be deleted if this instance was written to the repository.
+     * NOTE: Indices that are only found in {@code previous} but not in this instance are not included in the result.
+     *
+     * @param previous Previous {@code ShardGenerations}
+     * @return Map of obsolete shard index generations in indices that are still tracked by this instance
+     */
+    public Map<IndexId, Map<Integer, String>> obsoleteShardGenerations(ShardGenerations previous) {
+        final Map<IndexId, Map<Integer, String>> result = new HashMap<>();
+        previous.shardGenerations.forEach(((indexId, oldGens) -> {
+            final List<String> updatedGenerations = shardGenerations.get(indexId);
+            final Map<Integer, String> obsoleteShardIndices = new HashMap<>();
+            assert updatedGenerations != null
+                : "Index [" + indexId + "] present in previous shard generations, but missing from updated generations";
+            for (int i = 0; i < oldGens.size(); i++) {
```

Comment:
> What if we have an index with 5 primaries, snapshot it, delete the index, and create the same index with 1 primary, and then snapshot it.

That's not possible. It would have a different index uuid and would get a different repo uuid. The shard count for indices in the repo is unchangable as far as I can tell.
The only situaiton where it could be increasing is if the first snapshot to an index failed for the highest shardId(s) but succeeded for some lower id(s). It's never decreasing though.

Reply:
@ywelsch :man_facepalming: wow you're right I never noticed that (same name -> same indexUUID) for some reason. That's really unfortunate efficiency wise (not just in light of this change) ... though I guess it's an edge case.

I'll adjust he logic accordingly and add a test.

Reply:
Adjusted code to handle this 0f4b316edfc and added test for the scenario (had to adjust the consistency checks to understand this as well)

Reply:
@ywelsch `SnapshotsInProgress.Entry` now tracks the format of the snapshot 3a66c7966b3